ServiceCollectionExtensions

nsarris opened the pull request:
ServiceCollectionExtensions

nsarris commented on src/FluentValidation.DependencyInjectionExtensions/FluentValidation.DependencyInjectionExtensions.csproj:
```diff
@@ -31,8 +31,8 @@
-    <ItemGroup>
-        <PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.1.0" />
+    <ItemGroup>
+      <PackageReference Include="microsoft.extensions.dependencyinjection.abstractions" Version="2.2.0" />
```
This is the minimum dependency for this to work. I suggest you move anything AspNetCore related to another assembly so that this stays net framework compatible

JeremySkinner replied:
Please can this be downgraded to 2.1.0? FluentValidation supports .NET Core 2.1 as the lowest supported version, as this is the LTS release (2.2 is not LTS).

nsarris commented on src/FluentValidation/AssemblyScanner.cs:
```diff
@@ -69,6 +69,14 @@ public class AssemblyScanner : IEnumerable<AssemblyScanner.AssemblyScanResult> {
+		/// <summary>
+		/// Finds all the validators in the assembly containing the specified type.
+		/// </summary>
+		public static AssemblyScanner FindValidatorsInAssemblyContaining(Type type)
```
This is not used but I think it was missing anyway so I left it in

nsarris commented on src/FluentValidation.Tests.AspNetCore/ServiceCollectionExtensionTests.cs:
```diff
@@ -0,0 +1,30 @@
+﻿#if NETCOREAPP || NETSTANDARD
```
I initially placed this in the Tests project so I had to restrict it to core/standard. I forgot to remove it after I moved to to this project.

JeremySkinner commented on src/FluentValidation.DependencyInjectionExtensions/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,64 @@
+using System.Text;
+
+namespace FluentValidation
+{
```
Please can all braces be moved to the end of the preceding line for consistency with the rest of the project.

nsarris replied:
I think this is resolved

JeremySkinner replied:
Braces are still on new lines - please can they be moved to end of line? so `namespace FluentValidation { `

nsarris commented on src/FluentValidation.DependencyInjectionExtensions/ServiceCollectionExtensions.cs:
```diff
@@ -52,13 +52,21 @@ public static IServiceCollection AddValidatorsFromAssemblyContaining<T>(this ISe
 							implementationType: scanResult.ValidatorType,
 							lifetime: lifetime));
 
+			//Register as self
```
Added self registration so resolving by TValidator is also possible

JeremySkinner replied:
This introduces an inconsistency with the registration as part of the MVC `AddFluentValidation` call in `FluentValidation.AspNetCore`. The two need to be consistent, so a few options:

- `AddFluentValidation` registrations should pass through these new methods (means FV.AspNetCore has to take a dependency on FV.DI)
- The same change needs to be made manually to FV.AspNetCore
- Or this needs to be removed from here again (for now)

nsarris replied:
I went with option 1 because you already had the dependency there :)

JeremySkinner replied:
Sounds good, thanks!

nsarris commented:
Sorry about the tests failing, I initially used an existing validator, then when I switched projects (from Tests to AspnetCore.Tests) I created one inline but it was private and AssemblyScanner missed. I also added both kinds of resolutions in the test (TValidator and IValidator<T>). I'm pretty sure I run the test, but apparently I didn't...

JeremySkinner commented on src/FluentValidation.DependencyInjectionExtensions/FluentValidation.DependencyInjectionExtensions.csproj:
```diff
@@ -1,40 +1,40 @@
-<Project Sdk="Microsoft.NET.Sdk">
```
Latest commit seems to have reformatted the project file - can this be reverted? It makes it hard to review changes

nsarris commented on src/FluentValidation.DependencyInjectionExtensions/ServiceCollectionExtensions.cs:
```diff
@@ -8,12 +8,27 @@ namespace FluentValidation
+		/// <param name="assembly">The assemblies to scan</param>
+		/// <param name="lifetime">The lifetime of the validators. The default is transient</param>
+		/// <returns></returns>
+		public static IServiceCollection AddValidatorsFromAssemblies(this IServiceCollection services, IEnumerable<Assembly> assemblies, ServiceLifetime lifetime = ServiceLifetime.Transient)
```
Added a multiple assembly endpoint

nsarris commented on src/FluentValidation.AspNetCore/FluentValidationMvcExtensions.cs:
```diff
@@ -39,15 +39,13 @@ public static class FluentValidationMvcExtensions {
 
 			expr(config);
 
-			if (config.AssembliesToRegister.Count > 0) {
```
Removed RegisterTypes and replaced it with the call to AddValidatorsFromAssemblies extension method

JeremySkinner commented on src/FluentValidation.DependencyInjectionExtensions/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,87 @@
+namespace FluentValidation
+{
+	public static class ServiceCollectionExtensions
+	{
```
Move brace to end of previous line please

JeremySkinner commented on src/FluentValidation.DependencyInjectionExtensions/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,87 @@
+		/// <param name="lifetime">The lifetime of the validators. The default is transient</param>
+		/// <returns></returns>
+		private static IServiceCollection AddScanResult(this IServiceCollection services, AssemblyScanner.AssemblyScanResult scanResult, ServiceLifetime lifetime)
+		{
```
Move brace to end of previous line please.

JeremySkinner commented on src/FluentValidation.Tests.AspNetCore/ServiceCollectionExtensionTests.cs:
```diff
@@ -0,0 +1,29 @@
+using FluentValidation;
+
+namespace FluentValidation.Tests
+{
```
Move brace to end of previous line please

JeremySkinner commented on src/FluentValidation.Tests.AspNetCore/ServiceCollectionExtensionTests.cs:
```diff
@@ -0,0 +1,29 @@
+{
+
+	public class ServiceCollectionExtensionTests
+	{
```
Correct brace layout again please

JeremySkinner commented on src/FluentValidation.Tests.AspNetCore/ServiceCollectionExtensionTests.cs:
```diff
@@ -0,0 +1,29 @@
+
+		[Fact]
+		public void ServiceCollectionExtensionTests_should_pass_if_validator_is_resolved()
+		{
```
Braces again please

JeremySkinner commented on src/FluentValidation/AssemblyScanner.cs:
```diff
@@ -69,6 +69,14 @@ public class AssemblyScanner : IEnumerable<AssemblyScanner.AssemblyScanResult> {
+		/// Finds all the validators in the assembly containing the specified type.
+		/// </summary>
+		public static AssemblyScanner FindValidatorsInAssemblyContaining(Type type)
+		{
```
Move brace to end of previous line please (think this is the last one)

JeremySkinner commented:
Looking good, just the reformatting of the braces to match the rest of the project and then I think we're good

nsarris commented:
- Fixed braces (at first I read end of file for some reason and thought I fixed it by removing an empty line, doh!)
- Moved usings inside namespace for consistency
- Added license in library file for consistency
- Split and renamed tests for clarity

JeremySkinner commented:
Thanks, looks good. I’ll aim to get this merged in for the next release.

nsarris commented:
Ok what about the pluggable AssemblyScanner, I'm going to need to work on top of these changes. Maybe merge the pull request on a new branch so I can contiue pushing merge requests there?

JeremySkinner commented:
I've merged this in
